test/test_runner.cc

```diff
@@ -161,6 +163,13 @@ int TestRunner::runTests(int argc, char** argv) {
   // Reset all ENVOY_BUG counters.
   Envoy::Assert::resetEnvoyBugCountersForTest();
 
+  // Initialize log recording sink.
+  LogRecordingSink* recorder;
+  if (std::getenv("ENVOY_NO_LOG_SINK") == nullptr) {
+    recorder = absl::IgnoreLeak(new LogRecordingSink(Logger::Registry::getSink()));
+    Logger::Registry::getSink()->recorder_test_only_ = recorder;
+  }
```

Comment:
Could we make this a `InjectableSingleton` and we needn't to store it at the prod sink. We can inject it if necessary here and clear at the end of the main function. The lifetime of the record could also be managed by an unique pointer and we needn't to care the leak.

Reply:
We need a place to define the singleton. The test log macro header is unfortunately linked many times (as I found out - I can't quite explain why), so you'll end up with many singletons. Where do you recommend we put the singleton definition?

Reply:
Hmmm, it doesn't matter anyway. I can take a look after a while. You have did a great job. Only waiting a response from @abeyad. Then we can merge it!